wandb/sdk/wandb_artifacts.py

```diff
@@ -1175,61 +1153,122 @@ class WBArtifactHandler(StorageHandler):
-            os.unlink(link_creation_path)
-        os.symlink(os.path.abspath(link_target_path), link_creation_path)
+        if not os.path.islink(link_creation_path) and not os.path.exists(
+            link_creation_path
+        ):
+            os.symlink(link_target_path, link_creation_path)
 
         return link_creation_path
 
     def store_path(self, artifact, path, name=None, checksum=True, max_objects=None):
-        # Resolve the reference until the result is a concrete asset
-        # so that we don't have multiple hops.
-        artifact_id, artifact_file_path = WBArtifactHandler.parse_path(path)
-        target_artifact = PublicApi().artifact_from_id(util.hex_to_b64_id(artifact_id))
-        entry = target_artifact._manifest.get_entry_by_path(artifact_file_path)
-
-        while entry.ref is not None and urlparse(entry.ref).scheme == self._scheme:
-            artifact_id, artifact_file_path = WBArtifactHandler.parse_path(entry.ref)
-            target_artifact = PublicApi().artifact_from_id(
-                util.hex_to_b64_id(artifact_id)
+        """
+        Stores the file or directory at the given path within the specified artifact. In this
+        case we recursively resolve the reference until the result is a concrete asset so that 
+        we don't have multiple hops. TODO: This resolution could be done in the server for
+        performance improvements.
+
+        :param artifact: The artifact doing the storing
+        :param path: The path to store
+        :type path: str
+        :param name: If specified, the logical name that should map to `path`
+        :type name: str
+        :return: A list of manifest entries to store within the artifact
+        :rtype: list(ArtifactManifestEntry)
+        """
+
+        # Recursively resolve the reference until a concrete asset is found
+        while path is not None and urlparse(path).scheme == self._scheme:
+            artifact_id, artifact_file_path = WBArtifactHandler.parse_path(path)
+            target_artifact = PublicArtifact.from_id(
+                util.hex_to_b64_id(artifact_id), self.client
             )
+
+            # this should only have an effect if the user added the reference bu url string directly
+            # (in orther words they did not already load the artifact into ram.)
+            target_artifact._load_manifest()
+
             entry = target_artifact._manifest.get_entry_by_path(artifact_file_path)
+            path = entry.ref
 
+        # Create the path reference
         path = "wandb-artifact://{}/{}".format(
             util.b64_to_hex_id(target_artifact.id), str(entry.path)
         )
 
-        size = 0
+        # Return the new entry
         return [
             ArtifactManifestEntry(
-                name or os.path.basename(path), path, size=size, digest=path,
+                name or os.path.basename(path),
+                path,
+                size=0,
+                digest=path,
+                extra={
```

Comment:
Is this getting serialized into the actual artifact manifests that we store? We should make sure it's not. Manifest bytes are expensive (json parsing is expensive and we have to parse manifests a lot). The .ref field on the entry should be sufficient.

Reply:
This is removed in the incoming commit.